### ISSUE TITLE
Add option to refresh specific widget using Apple Script

### ISSUE THREAD

**Issue by rosenpin** [Jean-Tinland/simple-bar#205]
Currently, it's only possible to refresh the spaces widget/the data widget (although refreshing the data widget isn't even working for me) using ` osascript -e 'tell application id "tracesOf.Uebersicht" to refresh widget id "simple-bar-spaces-jsx"'` or `osascript -e 'tell application id "tracesOf.Uebersicht" to refresh widget id "simple-bar-data-jsx"' ` 

I suggest adding an option to refresh specific widgets, like the language widget for example. This way I could map my layout change key to change layout AND execute the apple script/any other script that triggers the language widget to update.  
For example something like:   
`osascript -e 'tell application id "tracesOf.Uebersicht" to refresh widget id "simple-bar-data-language-jsx"'`  
Though it seems like it wouldn't be as easy since spaces/data are a given, separating each widget would be painful.  
Another alternative could be to make spaces listen on some server, for example "127.0.0.1:41419", and allow users to do something like:   
`curl 127.0.0.1:41419/update/language`  
Essentially providing a basic REST API for updating/modifying widgets.  
This could also be useful for the Weather widget.  
You could theoretically expand this functionality to allow users add/remove widgets on the fly, for example:  
`curl 127.0.0.1:41419/show/spotify` to show the spotify widget when they want  
`curl 127.0.0.1:41419/hide/custom-1` to hide custom user widget

It could make simple bar much more customizable and powerful.  

I think the first thing that should be done is to add an option to update widgets on the fly, if you have other ideas to do that lets talk about them! :)

Edit:  
This could solve #199 using a remap (with better touch tool or even with the built in keyboard shortcuts manager, remap the layout change button to change layout AND update language widget)

**Comment by Jean-Tinland** [comment]
This is a great idea! I'm really liking the idea of a REST API doing the job.

I need to think about it and find an elegant way for implementing it. :)

**Comment by prurph** [comment]
I'd also like this—or at least for refreshing `simple-bar-data-jsx` to refresh the data widgets. My use case is that I have skhd configured to use various modes, and I'd like to display the active mode. This is easily done by reading a text file or env var, but I can't force simple bar to update when this occurs.

**Comment by Jean-Tinland** [comment]
I'm just starting working on this.

For now I have a git branch with a node server started on simple-bar's init script and a react context wrapping the whole app that is subscribing to a web socket opened by the server.

Once I have found a clean way to start this server and handle the websocket stably, I'll work on a notification system allowing to toggle or refresh a specific data widget on the fly.

If this is working, this server may allow to create a lot of new features like display notifications directly into simple-bar, create custom widget based on the focused app, etc...

I'll let you know when I have made sufficient progress :)

**Comment by prurph** [comment]
Awesome! Would you be interested in contributions for this? I have a fork of simplebar where I communicated over websockets with a separate API I was running, either using Python, Node, or Bun to push updates to simplebar. The API itself registered Yabai signals that used curl to hit the API. It worked well but basically gutted simplebar and pushed all communication with it through the API. I was okay with this because it was just for my personal use.

I would be interested in revisiting some of this if it would be welcome/helpful, but definitely don't want to be "too many cooks" if you already have a plan or work in progress.

**Comment by Jean-Tinland** [comment]
First, thank you for your response!

Yeah, of course I appreciate any help if I feel I can maintain the code base in the future! You can find the server repository [here](https://github.com/Jean-Tinland/simple-bar-server) and the WIP simple-bar branch [here](https://github.com/Jean-Tinland/simple-bar/tree/feature/simple-bar-server).

I had to separate it from simple-bar Übersicht widget as it was easier to implement this way.

My goal is to make this server an optional thing that user can install next to simple-bar so I think it should handle only things that simple-bar can't do by itself like interacting on the fly with the data widgets. The most added value would be the ability to refresh the user custom widgets on demand.

My current approach is to subscribe to an individual web socket in each data widget and handle messages coming from the HTTP server that can be call with simple `curl` commands.

For example :
```shell
# Force time widget refresh
curl http://localhost:7776/time/refresh

# Toggle time widget visibility
curl http://localhost:7776/time/toggle

# Force the user widget n°1 to refresh
curl http://localhost:7776/user-widget/refresh/1
```

The http server would be started with `pm2` and thus could be registered on startup but it could also be simply manually started when the user needs it.

These are my first thoughts, I'll look into what you have done in your fork. Is you API public so I can take a look too?

**Comment by prurph** [comment]
That sounds great! I mostly use simple-bar to display spaces and apps on
them, so my repo strictly has Yabai push events to simple-bar and doesn’t
have any endpoints to tell simple-bar to refresh or to update.

The code is here:
https://github.com/prurph/yabai-spaces-server.

I don’t have an up-to-date fork of simple-bar that works with this, but my
old fork is here:
https://github.com/prurph/simple-bar. The relevant parts is YabaiContext.ts
(
https://github.com/prurph/simple-bar/blob/main/lib/components/YabaiContext.jsx
).

**Comment by Jean-Tinland** [comment]
Thank you for that! There will indeed be some similarities between our approaches. I will look into what you have done.

First, I am finishing documenting all of simple-bar's features. This will allow me to de-clutter the README file and make room for highlighting new major features. After that, I will work on that subject and let you know when I have something that works.

**Comment by Jean-Tinland** [comment]
Continuing on the server implementation, I thought of what you said and I think I will also provide the option to refresh spaces & process parts of the bar like you did @prurph.

In this case, user will be responsible to add the right signals in their yabai config.

It should considerably improve the bar responsiveness.

**Comment by Jean-Tinland** [comment]
I just pushed the yabai (spaces, windows & displays) refresh in simple-bar-server. Documentation is available [here in simple-bar-server repo's readme file](https://github.com/Jean-Tinland/simple-bar-server#yabai).

@prurph using this method for all simple-bar refreshes greatly improve its responsiveness!

I batched every 20ms consecutive refreshes requested by `yabai` signals in order to prevent overlapping events to trigger multiple useless re-renders.
